### ISSUE TITLE
Conflict between webdataset implementation and OpenCLIP

### ISSUE THREAD

**Issue by wanghao14** [mlfoundations/open_clip#935]
I encountered the following error when finetuning OpenCLIP on my own data:
```
File "./src/training/data.py", line 281, in group_by_keys_nothrow
    fname, value = filesample["fname"], filesample["data"]
KeyError: 'fname'
```
This issue has been previously proposed in [Webdataset issue #384](https://github.com/webdataset/webdataset/issues/384). The problem appears to stem from a conflict in loading datasets in WebDataset format within OpenCLIP:
1. The function `tarfile_to_samples_nothrow` is added to the [pipeline](https://github.com/mlfoundations/open_clip/blob/main/src/open_clip_train/data.py#L375);
2. Within `tarfile_to_samples_nothrow`, `tar_file_expander ` is called to [iterate over opened tar file](https://github.com/mlfoundations/open_clip/blob/main/src/open_clip_train/data.py#L217). 
3. When the data in this thread is exhausted,  `{}` (the default value of `eof_value`)  is returned, as implemented in [webdataset](https://github.com/webdataset/webdataset/blob/main/webdataset/tariterators.py#L156).
4. When the given `{}`  is passed to `group_by_keys_nothrow` , the error is triggered.
As I am new to working with WebDatasets and not fully familiar with the underlying principles, I would appreciate any guidance or solutions you can provide.

**Comment by fartashf** [comment]
The fix is to pass `eof_value=None` to `tar_file_expander` on this line:
https://github.com/mlfoundations/open_clip/blob/fc5a37b72d705f760ebbc7915b84729816ed471f/src/open_clip_train/data.py#L217
This is a breaking change from webdataset and not sure if the right solution is to ask OpenCLIP to fix it or to ask webdataset to change the default value.

**Comment by rwightman** [comment]
hmm, that's a bit of a pain. Looks like webdataset might have been trying to fix an issue that I ran into and why I added this extra bit of code (nowthrow) in the first place, some webdatasets had colliding filenames across shards and would cause a crash. I think this might prevent that and other possible issues (completeing a sample with bits from two different shards)...

To fix I'd have to force using the latest webdataset and remove these hacks. And the painful part is verifying it all works :/

**Comment by wanghao14** [comment]
@rwightman Thank you for the detailed explanation regarding the cause of the issue. I really appreciate your efforts in maintaining the code.

Given the situation, do you think @fartashf's method could be a viable solution to retain your hacks while ensuring compatibility with the latest version of webdataset?

Thanks again for your hard work and dedication.

**Comment by rwightman** [comment]
For now I've constrained the webdataset version in pyproject.toml and requirements-training ... 'webdataset>=0.2.5,<=0.2.86' should be safe, I ran into another break caused by webdataset changes so feel this is the best approach for now

**Comment by wanghao14** [comment]
It appears that the changes in Webdataset have introduced a significant amount of additional work for user. I close this issue as it seems unrelated to the codebase of OpenCLIP. Moreover, I had transitioned from Webdataset to mosaicml-streaming for streaming training, which showed remarkably stable performance.

**Comment by rom1504** [comment]
If you want to PR mosaicml-streaming support into openclip it would be
welcome

**Comment by wanghao14** [comment]
I've been quite busy lately, but once I have some free time, I'd be delighted to submit a PR and contribute to this great project.

**Comment by vishaal27** [comment]
I also just faced this issue when training CLIP models where I had to cycle through a webdataset. My code was breaking with this same error when using `webdataset==0.2.100`, however the issue was resolved when I downgraded to `webdataset==0.2.86`.
